platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,343 @@
+The creation of a `Metrics` instance is not public api today.
+With the implementation of this proposal a `MetricsFactory` interface should be introduced that allows to create a `Metrics` instance.
+The `MetricsFactory` interface will be defined like this:
+
+```java
+interface MetricsFactory {
+    
+    @NonNull
+    default Metrics create() {
+        return create(Set.of());
+    }
+    
+    @NonNull
+    Metrics create(@NonNull Set<Label> defaultLabels);
+}
+```
+
+The definition of the api should follow our rules regarding services as defined at https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/docs/base/service-architecture/service-architecture.md
+
+##### Examples
+
+Here are some examples how to define a metric with labels:
+
+```java
+final Label transactionTypeLabel = new Label("transactionType", "fileUpload");
+final Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount").withLabels(transactionTypeLabel);
+
+final Metrics metrics = ...;
+final Counter counter = metrics.getOrCreate(config);
+
+counter.increment();
+```
+
+In prometheus the metric will be defined like this:
+
+```
+transactionsCategory.transactionCount{transactionType="fileUpload"}
+```
+
+#### Snapshot API
+
+The snapshot API will be extended to return the labels of a metric for a snapshot.
+By doing so any monitoring system that uses the snapshot API can use the labels.
+Today we only have 2 system that uses the snapshot API: the prometheus endpoint and the CSV exporter.
+Since no class of the snapshot API is public today (based on the internal usage of `nodeId`) we plan to move parts
+of the snapshot API to the public API.
+
+### Migration
+
+Next to extending the public API we need to migrate the current metrics to the new concept.
+
+#### Migration of the `category`, `name`, and `nodeId`
+
+The `nodeId` information of a metric will be migrated to labels.
+Today we create a prometheus label out of the `nodeId` internally in our prometheus endpoint implementation.
+The same conversion can be used to create a label for the `nodeId` in future but should be moved from the prometheus endpoint to the platform code.
+The key of a metric won't change and still be defined as the combination of the `category` and `name`.
+The unique identifier of the metric is the combination key and the labels.
+That is a breaking change since today the key is the unique identifier of a metric.
+Since we do not have an api today to add labels the migration should be minimalistic.
+
+#### Migration of the prometheus endpoint
+
+The `nodeId` of a platform is already added as a label today when using the prometheus endpoint.
+For that endpoint we just need to remove the hard coded `nodeId` label and add a general algorithm to add labels.
+
+#### Migration of the CSV exporter
+
+The CSV exporter use the `nodeId` today to create the file name.
+Since the CSV exporter is only used by the platform module we can continue with that approach.
```

Comment:
How are labels going to be used in the CSV-file?

Reply:
The nodeID label will be ignored and all other labels will be added to the name.

Reply:
But I assume it makes sense to make that more clear